kyo-core/shared/src/main/scala/kyo/Duration.scala

```diff
@@ -0,0 +1,136 @@
+    def fromJava(value: JavaDuration): Duration =
+        (value.toNanos: @annotation.switch) match
+            case 0                       => Zero
+            case n if n >= Long.MaxValue => Infinity
+            case n                       => n.nanos
+
+    def fromScala(value: ScalaDuration): Duration =
+        if value.isFinite then value.toNanos.nanos max Zero else Infinity
+
+    enum Units(val factor: Double):
+        case Nanos   extends Units(NANOS.getDuration.toNanos.toDouble)
+        case Micros  extends Units(MICROS.getDuration.toNanos.toDouble)
+        case Millis  extends Units(MILLIS.getDuration.toNanos.toDouble)
+        case Seconds extends Units(SECONDS.getDuration.toNanos.toDouble)
+        case Minutes extends Units(MINUTES.getDuration.toNanos.toDouble)
+        case Hours   extends Units(HOURS.getDuration.toNanos.toDouble)
+        case Days    extends Units(DAYS.getDuration.toNanos.toDouble)
+        case Weeks   extends Units(WEEKS.getDuration.toNanos.toDouble)
+        case Months  extends Units(MONTHS.getDuration.toNanos.toDouble)
+        case Years   extends Units(YEARS.getDuration.toNanos.toDouble)
+    end Units
+
+end Duration
+
+extension (value: Long)
+    inline def nanos: Duration   = value.asNanos
+    inline def micros: Duration  = value.as(Micros)
+    inline def millis: Duration  = value.as(Millis)
+    inline def seconds: Duration = value.as(Seconds)
+    inline def minutes: Duration = value.as(Minutes)
+    inline def hours: Duration   = value.as(Hours)
+    inline def days: Duration    = value.as(Days)
+    inline def weeks: Duration   = value.as(Weeks)
+    inline def months: Duration  = value.as(Months)
+    inline def years: Duration   = value.as(Years)
+end extension
+
+extension (self: Duration)
+
+    infix def >=(that: Duration): Boolean = self.gtEq(that)
+    infix def <=(that: Duration): Boolean = self.ltEq(that)
+    infix def >(that: Duration): Boolean  = self.gt(that)
+    infix def <(that: Duration): Boolean  = self.lt(that)
+    infix def ==(that: Duration): Boolean = self.eqEq(that)
+    infix def !=(that: Duration): Boolean = self.neEq(that)
+
+    inline def +(that: Duration): Duration = self.add(that)
+    inline def *(factor: Double): Duration = self.multiply(factor)
+
+    infix def max(that: Duration): Duration = self._max(that)
+    infix def min(that: Duration): Duration = self._min(that)
+
+    inline def toNanos: Long   = self._toNanos
+    inline def toMicros: Long  = self.to(Micros)
+    inline def toMillis: Long  = self.to(Millis)
+    inline def toSeconds: Long = self.to(Seconds)
+    inline def toMinutes: Long = self.to(Minutes)
+    inline def toHours: Long   = self.to(Hours)
+    inline def toDays: Long    = self.to(Days)
+    inline def toWeeks: Long   = self.to(Weeks)
+    inline def toMonths: Long  = self.to(Months)
+    inline def toYears: Long   = self.to(Years)
+
+    def toScala: ScalaDuration =
+        (self: @annotation.switch) match
```

Comment:
cool, I didn't know this annotation!

Reply:
it directly compiles `match` to a java switch. It's a good way to ensure performance, even if the compiler should do the right thing without it. It should fail compilation where the compilation is not possible.